cmd/cost/cost.go

```diff
@@ -0,0 +1,95 @@
+package cost
+
+import (
+	"fmt"
+	"github.com/openshift/osd-utils-cli/cmd/common"
+	awsprovider "github.com/openshift/osd-utils-cli/pkg/provider/aws"
+	"github.com/spf13/cobra"
+	"k8s.io/cli-runtime/pkg/genericclioptions"
+	cmdutil "k8s.io/kubectl/pkg/cmd/util"
+	"log"
+)
+
+// costCmd represents the cost command
+func NewCmdCost(streams genericclioptions.IOStreams) *cobra.Command {
+	opsCost = newCostOptions(streams)
+	costCmd := &cobra.Command{
+		Use:   "cost",
+		Short: "Cost Management related utilities",
+		Long: `The cost command allows for cost management on the AWS platform (other 
+platforms may be added in the future)`,
+		PersistentPreRun: func(cmd *cobra.Command, args []string) {
+			cmdutil.CheckErr(opsCost.complete(cmd, args))
+		},
+	}
+
+	//Set flags
+	costCmd.PersistentFlags().StringVarP(&opsCost.accessKeyID, "aws-access-key-id", "a", "", "AWS Access Key ID")
+	costCmd.PersistentFlags().StringVarP(&opsCost.secretAccessKey, "aws-secret-access-key", "x", "", "AWS Secret Access Key")
+	costCmd.PersistentFlags().StringVarP(&opsCost.profile, "aws-profile", "p", "", "specify AWS profile")
+	costCmd.PersistentFlags().StringVarP(&opsCost.configFile, "aws-config", "c", "", "specify AWS config file path")
+	costCmd.PersistentFlags().StringVarP(&opsCost.region, "aws-region", "g", common.DefaultRegion, "specify AWS region")
+
+	//Add commands
+	costCmd.AddCommand(newCmdGet(streams))
+	costCmd.AddCommand(newCmdReconcile(streams))
+	costCmd.AddCommand(newCmdCreate(streams))
+	costCmd.AddCommand(newCmdList(streams))
+
+	return costCmd
+}
+
+var opsCost *costOptions
+
+// costOptions defines the struct for running Cost command
+type costOptions struct {
+	// AWS config
+	accessKeyID     string
+	secretAccessKey string
+	configFile      string
+	profile         string
+	region          string
+
+	genericclioptions.IOStreams
+}
+
+func newCostOptions(streams genericclioptions.IOStreams) *costOptions {
+	return &costOptions{
+		IOStreams: streams,
+	}
+}
+
+//Initiate AWS clients for Organizations and Cost Explorer services using, if given, credentials in flags, else, credentials in the environment
+func (opsCost *costOptions) initAWSClients() (awsprovider.Client, error) {
+	//Initialize AWS clients
+	var (
+		awsClient awsprovider.Client
+		err       error
+	)
+	if opsCost.accessKeyID == "" && opsCost.secretAccessKey == "" {
```

Comment:
Is  there a way to make if the user provides profile, then its mutually exclusive with accessKeyId/SecretAccessKey? Then this could be simplified to 

if opsCost.profile != "" {

} else {

}

Reply:
I haven't found a way to do this yet

Reply:
https://github.com/openshift/osd-utils-cli/blob/master/cmd/federatedrole/apply.go#L70

Reply:
Ahh I think I misunderstood you. I'm looking at this further and I think there is a way. However, we cannot do 
```
if opsCost.profile != "" {

} else {

}
```
without refactoring the NewAwsClient and NewAwsClientWithInput functions, AFAIK. Since we need to call NewAwsClient in both cases where `opsCost.profile!=""` and `opsCost.profile == ""` (the latter occurs when all opsCost fields are null (with the exception of opsCost.region, which has a default value), in which we search for env vars), we can use your if/else format by moving implementing `sess := session.Must(session.NewSessionWithOptions(opt))` (https://github.com/openshift/osd-utils-cli/blob/72d26ea1f906232344c8ac6401a14bd904a5f447/pkg/provider/aws/client.go#L83) somewhere in NewAwsClientWithInput (https://github.com/openshift/osd-utils-cli/blob/72d26ea1f906232344c8ac6401a14bd904a5f447/pkg/provider/aws/client.go#L103).

I might need to change some of Ben's code as well if we want to do that. What are your thoughts?

Reply:
I see what your saying. What happens if a user doesn't provide access keys OR a profile?

Reply:
If that happens, the first if statement would execute to check for env vars. Namely, `NewAwsClient` runs, then `sess := session.Must(session.NewSessionWithOptions(opt))` runs. NewSessionWithOptions(opt) first loads the env vars (could be nil), then--if profile was given--sets the session's profile. NewSessionWithOptions exits and we're back at NewAwsClient. In our scenario, we'll then hit a "NoCredentialProviders" switch statement and return an error message. Program terminates with the message, `Error initializing AWS clients: Could not create AWS session: NoCredentialProviders: no valid providers in chain. Deprecated.`